src/tools/pub/custom-package-repositories.md

```diff
@@ -0,0 +1,119 @@
+[hosted packages](./dependencies#hosted-packages).
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
+
+In the example above `package:foo` will be fetched from
+`https://pub.example.com`. If authentication is required by this package
+repository, it can be specified using
+`dart pub token add https://pub.example.com`, which will prompt for the secret
+token.
+
+
+### Using multiple package repositories
+It is possible to fetch different dependencies from different package
+repositories, as the _hosted-url_ is specified for each dependency. As
+illustrated below.
+
+{% prettify yaml tag=pre+code %}
+dependencies:
+  # package retry is fetched from pub.dev (the default package repository)
+  retry: ^3.0.0
+  # package foo is fetched from https://pub.example.com
+  foo:
+    hosted: https://pub.example.com
+    version: ^1.4.0
+{% endprettify %}
+
+This enables you to keep private packages on a private package repository
+while staying up-to-date with latest public dependencies. However, conflicts can
+easily arise if package `retry` requires `meta` from pub.dev, and `foo` requires
+`meta` from `https://pub.example.com`. Thus, if mirroing packages into a private
+package repository is it frequently necessary to also mirror all dependencies
+and either update the `dependencies` section in section in each package, or
+override the default package repository.
+
+{{site.alert.note}}
+To ensure that public packages are usable to everyone, the official package
+repository [pub.dev](https://pub.dev) does not allow publication of packages
+with git-dependencies or hosted-dependencies with custom package repositories.
+
+However, such packages can be published to a custom package repository!
+{{site.alert.end}}
+
+
+### Overriding the default package repository
+The package repository to be used can be specified on a per-dependency basis,
+or the default package repository can be overriden with the
+[`PUB_HOSTED_URL`](./environment-variables) environment variable.
+
+This can be useful if you are mirroring all packages in a private package
+repository, or working in a restricted network environment using a mirror of
+[pub.dev](https://pub.dev).
+
+
+## Publishing to a custom package repository
+The `pub` tool also supports publishing packages to a private package
+repository, using the [`publish_to`](./pubspec#publish_to) property in
+`pubspec.yaml`. If working on a private package it is a good idea to specify
+this early in the development, so as to prevent accidental publication to
+[pub.dev](https://pub.dev). 
+
+To publish a package to `https://pub.example.com` you would write a
+`pubspec.yaml` as follows, and run `dart pub publish`.
+
+{% prettify yaml tag=pre+code %}
+name: mypkg
+version: 1.0.0
+# Ensures the package is published to https://pub.example.com
+publish_to: https://pub.example.com
+{% endprettify %}
```

Comment:
````suggestion
```
name: mypkg
version: 1.0.0
# Ensures the package is published to https://pub.example.com
publish_to: https://pub.example.com
```
````